napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,155 @@
+            if layer._is_async():
+                requests[layer] = layer._make_slice_request(dims)
+            else:
+                layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+        # # create task for slicing of each request/layer
+        # task = self._executor.submit(
+        #     self._slice_layers, requests
+        # )  # TODO: doesn't this submit all the llayer slicing under one thread?
+
+        # new_futures = {future: layer}
+        new_futures = []
+        with self.lock:
+            for layer, request in requests.items():
+                future = self._executor.submit(self._slice_layers, request)
+                future.add_done_callback(self._on_slice_done)
+                self._futures[future] = layer
+                new_futures.append(future)
+
+        # # construct dict of layers to layer task
+        # self._layers_to_task[tuple(requests.keys())] = task
+
+        # # once everything is complete, release the block
+        # task.add_done_callback(self._on_slice_done)
+
+        return new_futures
+
+    def shutdown(self) -> None:
+        """This should be called from the main thread when this is no longer needed."""
+        # TODO: kcp: doesn't this make this class instance null? Maybe we need a context manager to handle all instances off this class?
+        self._executor.shutdown()
+
+    def _slice_layers(self, requests: Dict) -> Dict:
+        """This can be called from the main or slicing thread.
+        Iterates throught a dictionary of request objects and call the slice
+        on each individual layer."""
+        return {
+            layer: layer._get_slice(request)
+            for layer, request in requests.items()
+        }
+
+    def _on_slice_done(self, task: Future[Dict]) -> None:
+        """This can be called from the main or slicing thread.
+        Release the thread.
+        This is the "done_callback" which is added to each task.
+        """
+        # TODO: kcp: I'm not happy with this ordering, it feels like we shouldn't remove it from the dict before
+        #       we actually call result, though I guess it doens't hurt anything to do so
+
+        # layers = self._find_layers_for_task(task)
+        # task_to_layers = {v: k for k, v in self._layers_to_task.items()}
+        # layers = task_to_layers.get(task, None)
+
+        layer = self._futures.get(task)
+
+        if not layer:
+            LOGGER.debug('Task not found')
+            return
+
+        # TODO: What if we submit layer1, then submit layer1 again? The first
+        # layer is complete and is running this. Meanwhile, the second layer
+        # has replaced the first layer in `_layers_to_task` dict. In this case,
+        # layer2 will be the one deleted here.
+        # remove task
+        # del self._layers_to_task[layers]
+        del self._futures[task]
+
+        if task.cancelled():
+            LOGGER.debug('Cancelled task')
+            return
+        result = task.result()
```

Comment:
FYI, at this point this call to `result()` won't wait. We know the future is done either because the task finished with an error, completed successfully, or was cancelled.

Reply:
That makes sense. Are you suggesting that something here isn't in alignment with that idea?